extensions/amp-story/1.0/parallax-fx.js

```diff
@@ -0,0 +1,352 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev, userAssert} from '../../../src/log';
+import {listen} from '../../../src/event-helper';
+import {mapRange, sum} from '../../../src/utils/math';
+import {setImportantStyles, setStyles} from '../../../src/style';
+import {toArray} from '../../../src/types';
+
+/** @enum {string} */
+export const ParallaxModes = {
+  /**
+   * Indicates that the last (nearest) layer will appear at the
+   * surface of the screen and all previous layers are behind it
+   * "inside" the screen.
+   */
+  DEPTH: 'depth',
+
+  /**
+   * Indicates that the first (farthest) layer will appear at the surface
+   * of the screen and all subsequent layers will appear as "popping out"
+   * of the screen.
+   */
+  POP_OUT: 'pop-out',
+
+  /**
+   * Indicates that the middle layer will appear at the surface of the
+   * screen with the previous (farther) layers appearing behind the screen
+   * and the subsequent (nearer) layers popping out of the screen.
+   */
+  CENTER: 'center',
+};
+
+const SMOOTHING_PTS = 4;
+const PERSPECTIVE = 1500;
+const DEFAULT_MODE = ParallaxModes.DEPTH;
+const DEFAULT_LAYER_SPACING = 1;
+const DEFAULT_FARTHEST_SCALE = 1.2;
+const DEFAULT_NEAREST_SCALE = 1;
+
+const MODE_ATTR = 'parallax-fx-mode';
+const LAYER_SPACING_ATTR = 'parallax-fx-layer-spacing';
+const NEAREST_SCALE_ATTR = 'parallax-fx-nearest-scale';
+const FARTHEST_SCALE_ATTR = 'parallax-fx-farthest-scale';
+const NO_PARALLAX_FX_ATTR = 'no-parallax-fx';
+
+/**
+ * Installs parallax handlers
+ */
+export class ParallaxService {
+  /**
+   * @param {!Window} global
+   * @param {!../../../src/service/vsync-impl.Vsync} vsync
+   * @param {!Element} story
+   */
+  constructor(global, vsync, story) {
+    /** @private {!Window} */
+    this.win_ = global;
+
+    /** @private @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {!Element} */
+    this.story_ = story;
+
+    /** @private {!Array<!ParallaxPage>} */
+    this.parallaxPages_ = [];
+
+    /** @private {number} */
+    this.middleX_ = 0;
+
+    /** @private {number} */
+    this.middleY_ = 0;
+
+    /** @private {Array} */
+    this.smoothingPointsX_ = [];
+
+    /** @private {Array} */
+    this.smoothingPointsY_ = [];
+
+    /** @private {string} */
+    this.storyMode_ = this.story_.getAttribute(MODE_ATTR);
+
+    /** @private {string} */
+    this.storyLayerSpacing_ = this.story_.getAttribute(LAYER_SPACING_ATTR);
+
+    /** @private {string} */
+    this.storyFarthestScale_ = this.story_.getAttribute(FARTHEST_SCALE_ATTR);
+
+    /** @private {string} */
+    this.storyNearestScale_ = this.story_.getAttribute(NEAREST_SCALE_ATTR);
+
+    this.init_();
```

Comment:
Constructors shouldn't "do" things, they should just construct objects. Should split construction and initialization, and have the caller initialize. It seems like initialization should also wait until the body has finished loading to ensure all pages are available for the `querySelector`.

Also not sure `querySelector` is the right way to go here. Can stories have pages added dynamically?

Reply:
Switched to a model that installs the effect on a per-page basis.